src/lib/protocols/mdns/Publisher.cpp

```diff
@@ -0,0 +1,171 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "Publisher.h"
+
+#include <inttypes.h>
+
+#include "lib/protocols/mdns/platform/Mdns.h"
+#include "lib/support/logging/CHIPLogging.h"
+#include "platform/CHIPDeviceBuildConfig.h"
+#include "platform/CHIPDeviceLayer.h"
+#include "support/CodeUtils.h"
+#include "support/ErrorStr.h"
+#include "support/RandUtils.h"
+
+namespace chip {
+namespace Protocols {
+namespace Mdns {
+
+CHIP_ERROR Publisher::Init()
+{
+    CHIP_ERROR error;
+
+    mUnprovisionedInstanceName = GetRandU64();
+    SuccessOrExit(error = ChipMdnsInit(HandleMdnsInit, HandleMdnsError, this));
+    SuccessOrExit(error = SetupHostname());
+exit:
+    return error;
+}
+
+void Publisher::HandleMdnsInit(void * context, CHIP_ERROR initError)
+{
+    Publisher * publisher = static_cast<Publisher *>(context);
+
+    if (initError == CHIP_NO_ERROR)
+    {
+        publisher->mInitialized = true;
+    }
+    else
+    {
+        ChipLogError(Discovery, "mDNS initialization failed with %s", chip::ErrorStr(initError));
+    }
+}
+
+void Publisher::HandleMdnsError(void * context, CHIP_ERROR initError)
+{
+    ChipLogError(Discovery, "mDNS error: %s", chip::ErrorStr(initError));
+}
+
+CHIP_ERROR Publisher::StartPublishDevice(chip::Inet::InterfaceId interface)
+{
+    CHIP_ERROR error;
+
+    if (chip::DeviceLayer::ConfigurationMgr().IsFullyProvisioned() != mIsPublishingProvisionedDevice)
+    {
+        SuccessOrExit(error = StopPublishDevice());
+        // Set hostname again in case the mac address changes when shifting from soft-AP to station
+        SuccessOrExit(error = SetupHostname());
+    }
+    mIsPublishingProvisionedDevice = chip::DeviceLayer::ConfigurationMgr().IsFullyProvisioned();
+
+    if (mIsPublishingProvisionedDevice)
+    {
+        error = PublishProvisionedDevice(interface);
+    }
+    else
+    {
+        error = PublishUnprovisionedDevice(interface);
+    }
+exit:
+    return error;
+}
+
+CHIP_ERROR Publisher::SetupHostname()
+{
+    uint8_t mac[6];
+    char hostname[13]; // Hostname will be the hex representation of mac.
+    CHIP_ERROR error;
+
+    SuccessOrExit(error = chip::DeviceLayer::ConfigurationMgr().GetPrimaryWiFiMACAddress(mac));
```

Comment:
This only works for Wifi, not thread. Thread EIDs are 64 bits. There should be no Wifi config directly read from common mDNS publisher, since for Ethernet based devices, there won't be a Wifi address. Please have a configuration method for MAC address (which may rotate over time, as well).

Reply:
The current platform api only provides Thread EID and wifi mac address. I prefer to update the paltform api to provide meta information about the underlying connectivity (Thread/wifi/Ethernet enabled or not).